stock_vertical_lift/models/stock_location.py

```diff
@@ -1,40 +1,131 @@
+                location.vertical_lift_kind = "view"
                 continue
             kind = tree.get(location.location_id.vertical_lift_kind)
             if kind:
                 location.vertical_lift_kind = kind
+
+    @api.depends(
+        "inverse_vertical_lift_shuttle_ids",
+        "location_id.vertical_lift_shuttle_id",
+    )
+    def _compute_vertical_lift_shuttle_id(self):
+        for location in self:
+            if location.inverse_vertical_lift_shuttle_ids:
+                # we have a unique constraint on the other side
+                assert len(location.inverse_vertical_lift_shuttle_ids) == 1
+                shuttle = location.inverse_vertical_lift_shuttle_ids
+            else:
+                shuttle = location.location_id.vertical_lift_shuttle_id
+            location.vertical_lift_shuttle_id = shuttle
+
+    def _fetch_vertical_lift_tray(self, cell_location=None):
+        """Send instructions to the vertical lift hardware
+
+        Private method, this is where the implementation actually happens.
+        Addons can add their instructions based on the hardware used for
+        this location.
+
+        The hardware used for a location can be found in:
+
+        ``self.vertical_lift_shuttle_id.hardware``
+
+        Each addon can implement its own mechanism depending of this value
+        and must call ``super``.
+
+        The method must send the command to the vertical lift to fetch / open
+        the tray. If a ``cell_location`` is passed and if the hardware supports
+        a way to show a cell (such as a laser pointer), it should send this
+        command as well.
+        """
+        if self.vertical_lift_shuttle_id.hardware == "simulation":
+            message = _("Opening tray {}.").format(self.name)
+            if cell_location:
+                message += _("<br/>Laser pointer on x{} y{}").format(
+                    cell_location.posx, cell_location.posy
+                )
+            self.env.user.notify_info(
+                message=message, title=_("Lift Simulation")
+            )
+
+    def fetch_vertical_lift_tray(self, cell_location=None):
+        """Send instructions to the vertical lift hardware
+
+        Public method to use for:
+        * fetch the vertical lift tray and presenting it to the operator
+          (physically)
+        * direct the laser pointer to the cell's location if set
+
+        Depending on the hardware, the laser pointer may not be implemented.
+
+        The actual implementation of the method goes in the private method
+        ``_fetch_vertical_lift_tray()``.
+        """
+        self.ensure_one()
+        if self.vertical_lift_kind == "cell":
+            if cell_location:
+                raise ValueError(
+                    "cell_location cannot be set when the location is "
+                    "a cell."
+                )
+            tray = self.location_id
+            tray.fetch_vertical_lift_tray(cell_location=self)
```

Comment:
@gurneyalex I don't think we need to have an additional method `fetch_vertical_lift_tray_show_cell()` since we can call `fetch_vertical_lift_tray()` on the cell directly

Reply:
let me know if you find it confusing though

Reply:
that's fine with me